Frametime v2

calebwang opened the pull request.

calebwang committed:
frametime validation now tracks missing frames, checks misaligned frames only if previous frame not missing

calebwang committed:
added test case for frame validation

calebwang committed:
added frametime column correction, tests

calebwang committed:
feature: guess units based on values

calebwang committed:
added midtimes feature

calebwang committed:
BF: caught a few bugs not covered in tests

calebwang committed:
new tests

calebwang committed:
improved csv parsing. a workaround for inconsistent delimiters

calebwang committed:
Modified from_csv, from_excel, to use new 6-column format. Modified to_csv to output correct data in new format.

calebwang committed:
Implemented new format for to_excel

calebwang committed:
minor touchups

cindeem commented:
When I ran the tests I get two errors (here is first)

Traceback (most recent call last):
  File "/home/jagust/cindeem/CODE/nipet/nipet/tests/test_frametime.py", line 57, in test_generate_protocol
    protocol = ft.generate_empty_protocol(4)
  File "/home/jagust/cindeem/CODE/nipet/nipet/frametime.py", line 194, in generate_empty_protocol
    outarray[0] = ['file number', 'expected frame', 'start time', 'duration', 'stop time', 'notes']
ValueError: cannot copy sequence with size 6 to array axis with dimension 4

which I can replicate:

myft = ft.FrameTime()
protocol = myft.generate_empty_protocol(4)

cindeem commented:
Im guessing the other error is related to getting timing data from ecat, you refer to a file "data/sample_frames.ecat" which doesnt exist....

Ill ignore this error for now.....and implement fix after merge

cindeem commented on nipet/frametime.py:
```diff
@@ -16,9 +16,64 @@ def _sec_to_min(seconds):
 def timestamp(filename):
     name, ext = splitext(filename)
-    return name + '-' + str(datetime.today()).replace(' ', '-') \
+    return name + '_' + str(datetime.today()).replace(' ', '-') \
```
Remember the issue with gzip extensions?  splitext('filename.nii.gz') only pulling .gz off?  this may also happen with these timing files...it is a rare corner case, but might be worth adding a check as users will commonly gzip "every" file in a directory

cindeem commented on nipet/frametime.py:
```diff
@@ -16,9 +16,64 @@ def _sec_to_min(seconds):
+    return name + '_' + str(datetime.today()).replace(' ', '-') \
                                              .split('.')[0] + ext
 
+def correct_data(data):
```
correct_data_order?

calebwang committed:
BF: Fixed generate_protocol

cindeem commented:
I just tried a round-trip
myft = ft.FrameTime().from_csv('tests/data/sample_frames.csv')
outarr = ft.generate_output(myft.data)
newft = ft.FrameTime().from_array(outarr, 'sec')

/home/jagust/cindeem/CODE/nipet/nipet/frametime.pyc in from_array(self, array, units)
    210             self._validate_frames()
    211         except FrameError:
--> 212             raise DataError('Bad data', self.data, 'array')
    213 
    214     def from_ecat(self, ecat_file, units=None):

DataError: 'Bad data' from 'array': array([[  1.,   1.,  15.,   0.,  15.],
       [  2.,   2.,  15.,  15.,  30.],
       [  3.,   4.,  15.,  45.,  60.],
       [  4.,   5.,  30.,  60.,  90.]])

So my question
1. The error message tells me I have bad data, but not why it is bad data, this could be improved
2. What is the outarray generated from ft.generate_output planned to be used for?

calebwang commented:
I've fixed/pushed the generate_protocol; it was just me being silly and forgetting to update the proper number of columns.

test_from_ecat is mostly just a reminder that from_ecat isn't implemented. 

I think I have something that will work for timestamps, which is to split twice if the first extension is in a set of predefined compression extensions, but I'll think about it some more. 

generate_output doesn't output a data array. it does some formatting on a data array and gives an array which has data to write to a file. mostly, this is just to insert blank lines on missing frames. I've changed the docstring to be less ambiguous.

cindeem commented:
Thanks for the quick changes, will discuss the user interface a little more next week
Given a FrameTime object, is there a fixed format I can use for running integration?  eg myft.data ?
Or will the structure of this array by very mutable.....maybe this is a additional feature to add to this module
myft.array_for_ga  (array for graphical analysis, fixed format array to use for integration?)

I could do this on my end, but seems to make more sense being in the ft module....thoughts?

calebwang commented:
myft.data/myft.get_data() should be fixed format (assuming nothing is assigned to it explicitly...)

format example:
[[1, 0, 300, 300], 
[3, 600, 300, 900], # drops missing frames, uses expected_frame (this can be changed to file_number easily)
[4, 900, 300, 1200]]

everything in a different format is computed based off of myft.data
